host/vm-manager/0008-Handle-proper-allocation-of-hugepages-VFs.patch

```diff
@@ -0,0 +1,326 @@
+-
+     return -1;
+ }
+ 
+@@ -386,14 +415,38 @@ bool VmBuilderQemu::SetupSriov(void) {
+     if (vf < 0)
+         return false;
+ 
++    int kVfUsed = vf;
++    if (kVfUsed == 0) {
++        LOG(error) << "Error: No VFs available";
++        return false;
++    }
++
+     emul_cmd_.append(" -device virtio-vga,max_outputs=" + ((numberOfDisplays > 0 ) ? std::to_string(numberOfDisplays) : "1") + ",blob=true"
+                     " -device vfio-pci,host=0000:00:02." + std::to_string(vf) +
+                     " -object memory-backend-memfd,hugetlb=on,id=mem_sriov,size=" + mem_size +
+                     " -machine memory-backend=mem_sriov");
+-
+     return true;
+ }
+ 
++void VmBuilderQemu::CleanupSriov() {
++    // Detect number of VFs
++    int numvf = ReadSysFile(kIntelGpuSriovNumVfs, std::ios_base::dec);
++
++    if (numvf != 0) {
++        // Check that all VFs are disabled
++        do_cleanup=true;
++        for (int avail=1; avail<=numvf; avail++ ) {
++	    std::string sd(kIntelGpuDevPath);
++            sd.replace(sd.end() - 2, sd.end(), std::to_string(avail) + "/enable");
++            int is_enabled = ReadSysFile(sd.c_str(), std::ios_base::dec);
++            if (is_enabled == 1) {
++                do_cleanup = false;
++                break;
++	    }
++	}
++    }
++}
++
+ void VmBuilderQemu::BuildExtraGuestPmCtrlCmd(void) {
+     std::string ex_cmd = cfg_.GetValue(kGroupExtra, kExtraPwrCtrlMultiOS);
+     if (ex_cmd == "true") {
+@@ -742,7 +795,12 @@ void VmBuilderQemu::BuildRpmbCmd(void) {
+     time(&rawtime);
+     localtime_r(&rawtime, &timeinfo);
+     strftime(t_buf, 80 , "%Y-%m-%d_%T", &timeinfo);
+-    std::string rpmb_sock = std::string(kRpmbSockPrefix) + t_buf;
++    std::string vm_name = cfg_.GetValue(kGroupGlob, kGlobName);
++    boost::trim(vm_name);
++    if (vm_name.empty()) {
++        return;
++    }
++    std::string rpmb_sock = std::string(kRpmbSockPrefix) + t_buf + vm_name;
+ 
+     if (!rpmb_bin.empty() && !rpmb_data.empty()) {
+         emul_cmd_.append(" -device virtio-serial,addr=1"
+@@ -927,14 +985,41 @@ void VmBuilderQemu::BuildVdiskCmd(void) {
+         " -device virtio-blk-pci,drive=disk1,bootindex=1");
+ }
+ 
++void VmBuilderQemu::SetupLockAcquire(void) {
++    setup_lock_fd = open(kSetupLock, O_RDWR | O_CREAT, S_IRWXU);
++    if (setup_lock_fd != -1) {
++        if (flock(setup_lock_fd, LOCK_EX) == -1) {
++            LOG(error)<<"acquire failed - ";
++            exit(-1);
++        }
++    } else {
++          LOG(error)<<"file open failed";
++    }
++}
++
++void VmBuilderQemu::SetupLockRelease(void) {
++    if (setup_lock_fd != -1) {
++        if (flock(setup_lock_fd, LOCK_UN) == -1) {
++            LOG(error)<<"release failed - ";
++            return;
++        }
++        close(setup_lock_fd);
++        setup_lock_fd = -1;
++    }
++}
++
+ bool VmBuilderQemu::BuildVmArgs(void) {
+     LOG(info) << "build qemu vm args";
++    SetupLockAcquire();
```

Comment:
Is the SRIOV VF assignment and hugepage allocation restricted within BuildVgpuCmd() function?
If yes, can we only acquire the lock only before calling BuildVgpuCmd() and releasing immediately after it returns?
We always want to avoid locking for too long a time, in case of any deadlock situation